third_party/amd/lib/TritonAMDGPUTransforms/OptimizeBufferOpPtr.cpp

```diff
@@ -0,0 +1,323 @@
+#include "triton/Analysis/Utility.h"
+#include "triton/Dialect/Triton/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/IR/Dialect.h"
+#include "triton/Dialect/TritonGPU/Transforms/Utility.h"
+
+#undef DEBUG_TYPE
+#define DEBUG_TYPE "tritonamdgpu-optimize-buffer-op-ptr"
+#define DBGS() (llvm::dbgs() << "[" DEBUG_TYPE "]: ")
+#define LDBG(X) LLVM_DEBUG(DBGS() << X << "\n")
+
+using ::mlir::LLVM::AMD::getVectorSize;
+using mlir::triton::AMD::ISAFamily;
+
+namespace ttg = mlir::triton::gpu;
+namespace tt = mlir::triton;
+namespace amdttg = mlir::triton::amdgpu;
+
+namespace mlir {
+
+#define GEN_PASS_DEF_TRITONAMDGPUOPTIMIZEBUFFEROPPTR
+#include "TritonAMDGPUTransforms/Passes.h.inc"
+
+namespace {
+// /*-----------------Base pointer increment optimization-------------------*/
+
+// Optimization tries to transfer increments from offsets to base pointer in
+// buffer operations:
+//
+// for ... (offsets = offsets_init):
+//   val = buffer_load basePtr [ offsets ]
+//   offsets_new = offsets + cst
+//   yield offsets_new
+// }
+//
+// transforms to:
+//
+// for ... (basePtr = basePtr_init):
+//   val = buffer_load basePtr [ offsets ]
+//   basePtr_new = basePtr + cst
+//   yield basePtr_new
+// }
+//
+// This lowers register consumption and reduces time spend for address
+// computation.
+struct AdvanceBasePointer : public OpRewritePattern<scf::ForOp> {
+  using OpRewritePattern::OpRewritePattern;
+
+  // Check if same value is stored in every element of tensor val
+  // Return true if all elelemts are equal, false if not or it was not able for
+  // proove it.
+  static bool isScalarizableValue(mlir::Value val) {
+    if (isa<IntegerType>(val.getType()))
+      return true;
+    auto defOp = val.getDefiningOp();
+    if (!defOp)
+      return false;
+    if (auto constantOp = dyn_cast<mlir::arith::ConstantOp>(defOp)) {
+      auto denseAttr =
+          dyn_cast<DenseIntElementsAttr>(constantOp.getValueAttr());
+      if (!denseAttr)
+        return false;
+      if (!denseAttr.isSplat())
+        return false;
+      return true;
+    } else if (auto splat = dyn_cast<triton::SplatOp>(defOp)) {
+      return isScalarizableValue(splat.getSrc());
+    }
+    return false;
+  }
+
+  // Generate a scalar value that is stored in provided RankedTensor val
+  static mlir::Value scalarizeValue(PatternRewriter &rewriter,
+                                    mlir::Value val) {
+    if (isa<mlir::IntegerType>(val.getType()))
+      return val;
+    auto defOp = val.getDefiningOp();
+    assert(defOp);
+    if (auto constantOp = dyn_cast<mlir::arith::ConstantOp>(defOp)) {
+      auto denseAttr =
+          dyn_cast<DenseIntElementsAttr>(constantOp.getValueAttr());
+      assert(denseAttr);
+      auto splatVal = denseAttr.getSplatValue<llvm::APInt>();
+      rewriter.setInsertionPoint(constantOp);
+      return rewriter.create<mlir::arith::ConstantIntOp>(
+          constantOp.getLoc(), denseAttr.getElementType(), splatVal);
```

Comment:
You can just use `matchPattern(val, m_Constant(&scalar))` and then create a Value out of `scalar`?